Topology aware scheduler: resources per numa approach

AlexeyPerevalov opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

It implements idea collected in this document: https://docs.google.com/presentation/d/1SR6XSIFsHkiWTws66LABpTiZaRwYk8IoRts4HQWE8Bc/edit#slide=id.g752682a7d2_68_129

Mostly focused on approach where some simplified version of TopologyManager is filtering pods at scheduler time. Design described in KEP(https://github.com/kubernetes/enhancements/pull/1858)

/kind feature

Commit:
Kube-scheduler NUMA Topology aware scheduling

This patch partly implements modified ideas proposed in:
https://docs.google.com/document/d/1XGTx6F8qgdq_zPvd87LAfhW1ObQ1McTDAoIWyokY88E
and
https://docs.google.com/document/d/1gPknVIOiu-c_fpLm53-jUAm-AGQQ8XC0hQYE7hq0L4c

The exact idea it implements is simplified version of TopologyManager in
the kube-scheduler.

In this version Informer is located in the plugin itself,
as an alternative it could be near the PodInformer, but in
this case additional abstraction layer for providing changes to
plugin is required like *SnapshotSharedLister*.

This implementation pass path to config to creat a clientset,
API allow to create clientset for NodeResourceTopology, based on
RESTClient from FrameworkHandler.Discovery().RESTClient(), but
event after adding Schema, original client can't deserialise
NodeResourceTopology

CRD is described in this document:
https://docs.google.com/document/d/12kj3fK8boNuPNqob6F_pPU9ZTaNEnPGaXEooW1Cilwg/edit

TODO: move from github.com/AlexeyPerevalov to repository where daemon
should live.

Signed-off-by: Alexey Perevalov <alexey.perevalov@huawei.com>

Commit:
Introduce TopologyPolicy string constants

This constants contain topology policy information on the
worker node.

It should be used in resource with topology exporting daemon daemon,
in topology manager itself and in kube-scheduler plugin (TopologyMatch)

TODO move constants from topology_manager to this one

Signed-off-by: Alexey Perevalov <alexey.perevalov@huawei.com>

cbf123 commented on pkg/kubelet/nodestatus/setters.go:
```diff
@@ -52,6 +53,20 @@ const (
+	}
+
+	for node, cpuSet := range info {
+		numaKey := fmt.Sprintf("numa%d-cpu", node)
```
Rather than prepending with the numa information, I wonder if it might not make sense to have it be the suffix?  There is precedent for this in the linux kernel, which uses a path like "/sys/bus/pci/devices/0000:b3:00.0/numa_node" to report the NUMA node of a PCI device.  This would also let you get the "un-numified" device name by simply dropping the last path segment rather than needing to remove a string from the middle of a path like you have in your suggested numa directory.

AlexeyPerevalov replied:
yes, this approach has number of drawbacks. And we decided to move to CRD https://docs.google.com/document/d/12kj3fK8boNuPNqob6F_pPU9ZTaNEnPGaXEooW1Cilwg/edit.

swatisehgal committed:
Support corner case with memory

 - check for memory
 - code style fixes

Signed-off-by: Swati Sehgal <swsehgal@redhat.com>
Signed-off-by: Alexey Perevalov <alexey.perevalov@huawei.com>

Commit:
Add unit test for TopologyMatch plugin

It tests only Filter function, but doesn't test CRDInformer's handlers.

Signed-off-by: Alexey Perevalov <alexey.perevalov@huawei.com>

Commit:
QoS based corner cases for TopologyMatch

This commit as removes exporting of NodeTopologyMap type.
Also it more strictly keeps principle of marking NUMA node as
possible (bit set) in corner cases.

Signed-off-by: Alexey Perevalov <alexey.perevalov@huawei.com>

swatisehgal commented on pkg/scheduler/framework/plugins/noderesources/topology_match.go:
```diff
@@ -0,0 +1,316 @@
+}
+
+func (tm TopologyMatch) PolicyFilter(pod *v1.Pod, zoneMap topologyv1alpha1.ZoneMap) *framework.Status {
+	if v1qos.GetPodQOS(pod) != v1.PodQOSGuaranteed {
```
@AlexeyPerevalov Was testing out this latest implementation and was thinking about pod with resources as can be seen below where resource request= limit. In such cases, v1qos.GetPodQOS returns Burstable QoS (as memory is missing I guess?). 
```
resources:
      limits:
        cpu: 1
        example.com/deviceA: 1
        example.com/deviceB: 1
      requests:
        cpu: 1
        example.com/deviceA: 1
        example.com/deviceB: 1
```
This check here completely disregards such pods as they are not in the guaranteed QoS class, we might want to address this at least for the time being where there is no way to align memory based on topology?

AlexeyPerevalov replied:
Memory in requests is necessary for guaranteed QoS. But in topology manager makes alignment for all QoS, and such behaviour was here too. Looks like it was missed during last rebase/factoring out. I'll fix it.

swatisehgal committed:
NodeResourceTopology API in staging

Signed-off-by: Swati Sehgal <swsehgal@redhat.com>
Signed-off-by: Alexey Perevalov <alexey.perevalov@huawei.com>

swatisehgal committed:
Updates based on review
 - regenrated code
 - removed  custom-boilerplate
 - updated go.mod
 - regenerated the code and updated Bazel

Signed-off-by: Swati Sehgal <swsehgal@redhat.com>

Commit:
FeatureGate for NodeResourceTopologyMatch plugin

This commit follows approach when scheduler plugin is not
registered and as result even not instantiated.
Another approach to instantiate plugin, but do not start
Listers/Informers, but  in looks like none optimal.

Signed-off-by: Alexey Perevalov <alexey.perevalov@huawei.com>

ehashman commented:
Needs rebase.

AlexeyPerevalov commented:
/close
in favor to out-of-tree plugin
